File: Zitate/Utilities/Validators.swift
//
//  Validators.swift
//  Zitate
//
//  Created on 2026-03-14.
//

import Foundation

enum Validators {

    /// Validate text entry
    static func validateEntryText(_ text: String?) -> ValidationResult {
        guard let text = text, !text.isEmpty else {
            return .invalid("Text cannot be empty")
        }

        if text.count < Constants.minTextLength {
            return .invalid("Text must be at least \(Constants.minTextLength) character")
        }

        if text.count > Constants.maxTextLength {
            return .invalid("Text cannot exceed \(Constants.maxTextLength) characters")
        }

        return .valid
    }

    /// Validate author name
    static func validateAuthorName(_ name: String?) -> ValidationResult {
        guard let name = name, !name.isEmpty else {
            return .invalid("Author name cannot be empty")
        }

        if name.count > Constants.maxAuthorNameLength {
            return .invalid("Author name cannot exceed \(Constants.maxAuthorNameLength) characters")
        }

        return .valid
    }

    /// Validate label name
    static func validateLabelName(_ name: String?) -> ValidationResult {
        guard let name = name, !name.isEmpty else {
            return .invalid("Label name cannot be empty")
        }

        if name.count > Constants.maxLabelNameLength {
            return .invalid("Label name cannot exceed \(Constants.maxLabelNameLength) characters")
        }

        // Check for disallowed characters
        if name.contains(",") || name.contains(";") {
            return .invalid("Label name cannot contain commas or semicolons")
        }

        return .valid
    }

    /// Validate URL
    static func validateURL(_ urlString: String?) -> ValidationResult {
        guard let urlString = urlString, !urlString.isEmpty else {
            return .valid // URL is optional
        }

        guard URL(string: urlString) != nil else {
            return .invalid("Invalid URL format")
        }

        return .valid
    }

    /// Validate folder name
    static func validateFolderName(_ name: String?) -> ValidationResult {
        guard let name = name, !name.isEmpty else {
            return .invalid("Folder name cannot be empty")
        }

        if name.count > Constants.maxFolderNameLength {
            return .invalid("Folder name cannot exceed \(Constants.maxFolderNameLength) characters")
        }

        return .valid
    }
}

enum ValidationResult: Equatable {
    case valid
    case invalid(String)

    var isValid: Bool {
        if case .valid = self {
            return true
        }
        return false
    }

    var errorMessage: String? {
        if case .invalid(let message) = self {
            return message
        }
        return nil
    }
}

File: Zitate.web/src/utils/validators.ts
/**
 * Validation utilities for Zitate
 */

export type ValidationResult =
  | { isValid: true }
  | { isValid: false; error: string };

/**
 * Constants for validation
 */
export const VALIDATION_LIMITS = {
  TEXT_MIN: 1,
  TEXT_MAX: 10000,
  AUTHOR_NAME_MAX: 200,
  LABEL_NAME_MAX: 50,
  FOLDER_NAME_MAX: 100,
  LOCATION_MAX: 200,
} as const;

/**
 * Validate entry text
 */
export function validateEntryText(text: string | undefined | null): ValidationResult {
  if (!text || text.trim().length === 0) {
    return { isValid: false, error: 'Text cannot be empty' };
  }

  if (text.length < VALIDATION_LIMITS.TEXT_MIN) {
    return { isValid: false, error: `Text must be at least ${VALIDATION_LIMITS.TEXT_MIN} character` };
  }

  if (text.length > VALIDATION_LIMITS.TEXT_MAX) {
    return { isValid: false, error: `Text cannot exceed ${VALIDATION_LIMITS.TEXT_MAX} characters` };
  }

  return { isValid: true };
}

/**
 * Validate author name
 */
export function validateAuthorName(name: string | undefined | null): ValidationResult {
  if (!name || name.trim().length === 0) {
    return { isValid: false, error: 'Author name cannot be empty' };
  }

  if (name.length > VALIDATION_LIMITS.AUTHOR_NAME_MAX) {
    return { isValid: false, error: `Author name cannot exceed ${VALIDATION_LIMITS.AUTHOR_NAME_MAX} characters` };
  }

  return { isValid: true };
}

/**
 * Validate label name
 */
export function validateLabelName(name: string | undefined | null): ValidationResult {
  if (!name || name.trim().length === 0) {
    return { isValid: false, error: 'Label name cannot be empty' };
  }

  if (name.length > VALIDATION_LIMITS.LABEL_NAME_MAX) {
    return { isValid: false, error: `Label name cannot exceed ${VALIDATION_LIMITS.LABEL_NAME_MAX} characters` };
  }

  // Check for disallowed characters
  if (name.includes(',') || name.includes(';')) {
    return { isValid: false, error: 'Label name cannot contain commas or semicolons' };
  }

  return { isValid: true };
}

/**
 * Validate URL format
 */
export function validateURL(url: string | undefined | null): ValidationResult {
  if (!url || url.trim().length === 0) {
    return { isValid: true }; // URL is optional
  }

  try {
    new URL(url);
    return { isValid: true };
  } catch {
    return { isValid: false, error: 'Invalid URL format' };
  }
}

/**
 * Validate folder name
 */
export function validateFolderName(name: string | undefined | null): ValidationResult {
  if (!name || name.trim().length === 0) {
    return { isValid: false, error: 'Folder name cannot be empty' };
  }

  if (name.length > VALIDATION_LIMITS.FOLDER_NAME_MAX) {
    return { isValid: false, error: `Folder name cannot exceed ${VALIDATION_LIMITS.FOLDER_NAME_MAX} characters` };
  }

  return { isValid: true };
}

/**
 * Normalize label name (lowercase)
 */
export function normalizeLabelName(name: string): string {
  return name.trim().toLowerCase();
}

/**
 * Format label for display (capitalize first letter)
 */
export function formatLabelForDisplay(name: string): string {
  return name.charAt(0).toUpperCase() + name.slice(1);
}
